File: Cloud Computing Application/MP9/java/src/main/java/MP3_PartD.java
import scala.Tuple2;

import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.types.StringType;
import org.apache.spark.sql.types.IntegerType;

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.regex.Pattern;
//import java.util.function.Function;

public final class MP3_PartD {

  public static void main(String[] args) throws Exception {
    SparkSession spark = SparkSession
      .builder()
      .appName("MP3")
      .getOrCreate();
    JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());
    SQLContext sqlContext = new SQLContext(sc);
    /*
     * 1. Setup (10 points): Download the gbook file and write a function 
     * to load it in an RDD & DataFrame
     */
    
    // RDD API
    // Columns: 0: place (string), 1: count1 (int), 2: count2 (int), 3: count3 (int)


    // Spark SQL - DataSet API


    /*
     * 4. MapReduce (10 points): List the three most frequent 'word' with their count 
     * of appearances
     */

    // Dataset/Spark SQL API


    spark.stop();
    sc.stop();
  }
}

File: Cloud Computing Application/MP9/python/MP3_PartD.py
from pyspark import SparkContext, SQLContext
from pyspark.sql.types import StructType
from pyspark.sql.types import StructField
from pyspark.sql.types import StringType, IntegerType

sc = SparkContext()
sqlContext = SQLContext(sc)

####
# 1. Setup (10 points): Download the gbook file and write a function to load it in an RDD & DataFrame
####

lines = sc.textFile("gbooks", 10)
parts = lines.map(lambda l: l.split())
gbooks = parts.map(lambda p: (p[0], int(p[1]), int(p[2]), int(p[3])))

# RDD API
# Columns:
# 0: place (string), 1: count1 (int), 2: count2 (int), 3: count3 (int)

fields = [
    StructField("word", StringType()),
    StructField("count1", IntegerType()),
    StructField("count2", IntegerType()),
    StructField("count3", IntegerType())
]
schema = StructType(fields)

# Spark SQL - DataFrame API

df = sqlContext.createDataFrame(gbooks, schema)

####
# 4. MapReduce (10 points): List the three most frequent 'word' with their count of appearances
####

df.createOrReplaceTempView("gbooks")

# Spark SQL

sqlDF = sqlContext.sql("SELECT word, count(1) FROM gbooks GROUP BY word ORDER BY count(1) DESC")
sqlDF.show(n=3)

# There are 18 items with count = 425, so could be different 
# +---------+--------+
# |     word|count(1)|
# +---------+--------+
# |  all_DET|     425|
# | are_VERB|     425|
# |about_ADP|     425|
# +---------+--------+

